usr/local/bin/um-setup

```diff
@@ -2,45 +2,132 @@
-  echo "Usage: $(basename $0) [GCOM source tree]"
+function usage {
+  # Print command line options
+  echo 'Usage: um-setup [-g gcom_path] [-s shumlib_path]'
+  echo
+  echo 'Options:'
+  echo '  -g path    GCOM source location (working copy or repository location)'
+  echo '  -h         Show this help and exit'
+  echo '  -s path    shumlib source location (working copy or repository location)'
+}
+
+function ereport {
+  # Print an error message, print usage, then exit (non-zero)
+  echo ${1:-"Unknown error"}
+  usage
   exit 1
-fi
+}
 
-src=fcm:gcom.x/trunk@head
-if [[ -n "$1" ]]; then
-  src=$1
-fi
+function process_args {
+  # Parse and process the command line arguments
+  if [ $# -gt 4 ]; then
+    ereport "Too many arguments"
+    exit 1
+  fi
 
-tmpdir=$(mktemp -d)
+  while [ $# -ge 1 ] ; do
+    case "$1" in
+      -g) shift
+          if [ -n "$1" ]; then
+            gcom_src=$1
+          else
+            ereport "GCOM source not provided"
+          fi
+          ;;
+      -s) shift
+          if [ -n "$1" ]; then
+            shumlib_src=$1
+          else
+            ereport "shumlib source not provided"
+          fi
+          ;;
+      -h) usage
+          exit 0
+          ;;
+       *) ereport "Unrecognised argument: $1"
+          ;;
+      esac
+      shift
+  done
+}
 
-if [[ $src =~ ^fcm:|^svn:|file: ]]; then
-  # A URL of some kind
-  # Check the location is valid and get the Last Changed Rev
-  rev=$(fcm info $src | awk '/Last Changed Rev/ {print $NF}')
-  if [[ -z $rev ]]; then
-    echo "Failed to find a source tree at $src"
-    exit 1
+function get_source_tree {
+  # Put a copy of the src $1 in tmpdir $2,
+  # whether a working copy or repository location,
+  # and return either the revision for a repos, or 'wc' for a working copy.
+  # This is because no check is made whether the wc contains local changes.
+  src=$1
+  tmpdir=$2
+
+  if [[ $src =~ ^fcm:|^svn:|file: ]]; then
+    # A URL of some kind
+    # Check the location is valid and get the Last Changed Rev
+    rev=$(fcm info $src | awk '/Last Changed Rev/ {print $NF}')
+    if [[ -z $rev ]]; then
+      echo "Failed to find a source tree at $src"
+      exit 1
+    fi
+    vn="r$rev"
+    fcm co -q $src $tmpdir/wc || { echo "Failed to checkout $src"; exit 1; }
+  else
+    # Assume working copy
+    if [[ ! -d $src ]]; then
+      echo "$src is not a recognised repository location or a valid directory"
+      exit 1
+    fi
+    vn=wc
+    cp -r $src $tmpdir/wc || { echo "Copy of $src to $tmpdir failed."; exit 1; }
   fi
-  vn="r$rev"
-  fcm co -q $src $tmpdir/gcom_wc
-else
-  # Assume working copy
-  if [[ ! -d $src ]]; then
-    echo "$src is not a recognised repository location or a valid directory"
-    exit 1
+}
+
+function create_keyword {
+  # Create a (primary) keyword in keyword.cfg if it doesn't exist.
+  keyword=$1
+  keypath=$2
+  keyfile=$HOME/.metomi/fcm/keyword.cfg
+  if ! grep -qL "location{primary}\[$keyword\]" $keyfile; then
```

Comment:
This doesn't check that the keyword is set to the correct path, I think, just that it exists? (Unlike the rose.conf rose-ana database test later which does check that it's true).